### PR TITLE
Don't create unnecessary render layers.

### PR DESCRIPTION
**Pull request by grokys**
When doing a deferred render, if a control has an `Opacity` < 1 but has no visual children, instead of creating a new render layer, bake the opacity into the draw operations.

The main thing needed to get this working was better tools for manipulating brushes.

This reduces the number of render layers created on the Calendar page of the ControlCatalog from 357 to 54 and the FPS on that page from 6 to 60 on my machine. There's still more to be done here, but this is a decent first step.

tl;dr: it turns `<Rectangle Fill="Red" Opacity="0.5"/>` into `<Rectangle Fill="#80ff0000"/>`

### CODE REVIEW

**Comment by kekekeks** [comment]
default
https://www.html5rocks.com/en/tutorials/speed/layers/

That's a bit old, but we might use the same criterias.

**Comment by grokys** [comment]
Yeah we will probably need to modify the heuristics for when we create layers going forwards.

Thing is, that for opacity even if we don't create a render layer in the deferred renderer, at least in D2D, pushing opacity requires pushing a D2D layer anyway. This fix here gets around needing to do that for simple primitives such as `Rectangle` by baking the transparency into the draw operation, as suggested here: https://msdn.microsoft.com/en-us/library/windows/desktop/dd756654(v=vs.85).aspx#alternatives_to_layers

**Comment by kekekeks** [comment]
>at least in D2D, pushing opacity requires pushing a D2D layer anyway

Internal D2D layers are stacked composited in-place and reused, so one won't have more than 10 at the same time for a reasonable control tree depth. Because of that D2D's layers are way more memory efficient than our layers that are composited when everything is done. Basically D2D's approach trades performance for memory while ours trades memory for performance. We need to wisely combine these approaches.

**Comment by kekekeks** [comment]
>Lastly, if you want to apply opacity to a single primitive, you should multiply the opacity into the into the brush color and then render the primitive. You do not need a layer or an opacity mask bitmap.

That's how our entire Skia backend works with opacity. It basically maintains the current "global" opacity value and multiplies everything on it before rendering.

**Comment by grokys** [comment]
> That's how our entire Skia backend works with opacity. It basically maintains the current "global" opacity value and multiplies everything on it before rendering.

Yeah I was thinking about this - if I remember correctly there is/was some sort of problem with this, is that correct? Has that been fixed?

But more generally, do you think this is a good idea or do you think we should be doing something different?

**Comment by kekekeks** [comment]
>there is/was some sort of problem with this, is that correct

Some parts of the code (bitmaps, text rendering) were ignoring that opacity setting. There also was a problem with subpixel text rendering that has to be disabled in case of text with opacity. I think it works fine now.

**Comment by kekekeks** [comment]
>do you think we should be doing something different

Layer creation and composition is expensive, but on the other hard rendering stuff with opacity might be expensive too. I think we shouldn't create D2D layers when possible and use opacity multiplication approach in DrawingContext code when possible and improve layer logic in DeferredRenderer.

**Comment by grokys** [comment]
So just to clarify, do you think we should merge this PR or take another course of action?

**Comment by kekekeks** [comment]
Sorry, I haven't actually taken a look at the code yet :(

**Comment by grokys** [comment]
Having thought about this some more, I think you're probably right that we should just create new render layers for controls that are animating and have children, so I'm going to close this.
